### PR TITLE
Dev #T238: Enable duplication of survey including pictures

### PR DESCRIPTION
**Pull request by gabrieljenik**

### CODE REVIEW

File: application/controllers/SurveyAdministrationController.php [hunk 1]
```diff
@@ -2183,7 +2184,9 @@ public function actionCopy()
                     );
                 }
             } elseif ($action == 'copysurvey' && !$aData['bFailed']) {
-                $aImportResults = XMLImportSurvey('', $copysurveydata, $sNewSurveyName, sanitize_int(App()->request->getParam('copysurveyid')), (Yii::app()->request->getPost('copysurveytranslinksfields') == '1'));
+                $copyResources = Yii::app()->request->getPost('copysurveytranslinksfields') == '1';
```
Comment (Shnoulle) [inline application/controllers/SurveyAdministrationController.php:2187]: `$copyResources = Yii::app()->request->getPost('copysurveytranslinksfields'); `

or 

`$copyResources = boolval(Yii::app()->request->getPost('copysurveytranslinksfields')); `

File: application/helpers/admin/import_helper.php [hunk 2]
```diff
@@ -1645,8 +1645,10 @@ function XMLImportSurvey($sFullFilePath, $sXMLdata = null, $sNewSurveyName = nul
             }
             unset($insertdata['id']);
             // now translate any links
-            $insertdata['question'] = isset($insertdata['question']) ? translateLinks('survey', $iOldSID, $iNewSID, $insertdata['question']) : '';
-            $insertdata['help'] = isset($insertdata['help']) ? translateLinks('survey', $iOldSID, $iNewSID, $insertdata['help']) : '';
+            if ($bTranslateInsertansTags) {
```
Comment (Shnoulle) [inline application/helpers/admin/import_helper.php:1648]: This is another fix here ?

File: application/models/services/CopySurveyResources.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+     * @param string $targetDir
+     *
+     * @return array An array of failed and copied files/directories
+     */
+    private function copyDirectory($sourceDir, $targetDir): array
+    {
+        $copiedFilesInfo = [];
+        $errorFilesInfo = [];
+
+        if (file_exists($sourceDir)) {
```
Comment (Shnoulle) [inline application/models/services/CopySurveyResources.php:42]: Why not is_dir ?

File: application/models/services/CopySurveyResources.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                        $subExtractdir = $sourceDir . "/" . $direntry;
+                        $subDestdir = $targetDir . "/" . $direntry;
+                        list($subImportedFilesInfo, $subErrorFilesInfo) = $this->copyDirectory($subExtractdir, $subDestdir);
+                        $copiedFilesInfo = array_merge($copiedFilesInfo, $subImportedFilesInfo);
+                        $errorFilesInfo = array_merge($errorFilesInfo, $subErrorFilesInfo);
+                    }
+                }
+            }
+        }
+        return [$copiedFilesInfo, $errorFilesInfo];
```
Comment (Shnoulle) [inline application/models/services/CopySurveyResources.php:82]: :+1:

File: application/views/surveyAdministration/tabCopy_view.php [hunk 5]
```diff
@@ -61,7 +61,7 @@
                 <div class="ls-flex-column col-md-4">
                     <!-- Convert resource links -->
                     <div class="form-group">
-                        <label class=" control-label" for='copysurveytranslinksfields'><?php echo  eT("Convert resource links and expression fields?"); ?> </label>
+                        <label class=" control-label" for='copysurveytranslinksfields'><?php echo  eT("Copy survey resource files and adapt links"); ?> </label>
```
Comment (Shnoulle) [inline application/views/surveyAdministration/tabCopy_view.php:64]: No way to convert link *without** copy directory ? Why not, but maybe need an extra option ?

**Review by Shnoulle — CHANGES_REQUESTED** [review]
Move to is_dir ? No ?

File: application/controllers/SurveyAdministrationController.php (hunk 1 above)
Comment (gabrieljenik) [reply application/controllers/SurveyAdministrationController.php:2187]: Thos could be a very subjective discussion... but why any of those options is better?

Comment (Shnoulle) [reply application/controllers/SurveyAdministrationController.php:2187]: == 1 is clearly uneeded.
and i think scrutinizer send a warning about usage of == and not === 

We check a boolean

Comment (gabrieljenik) [reply application/controllers/SurveyAdministrationController.php:2187]: I prefer to leave it like that. It was like these before. There are many other places on the current file where this is done like that. Prefer to leave it like that for consistency (by the way, I think is a bit safer as it is a tighter check)

**Comment by gabrieljenik** [comment]
> Move to is_dir ? No ?

I think this change request can be dismissed, right?

**Review by Shnoulle — CHANGES_REQUESTED** [review]
See new request 
1. Seems there are a fix , then need to be in commit "Fixed issue : link updated when copy survey in all condition"
2. Survey without file (no sourcedir) are not an error. I have a lot of surey without related survey/upload dir

File: application/models/services/CopySurveyResources.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                return [$copiedFilesInfo, $errorFilesInfo];
+            }
+
+            if (!is_dir($targetDir) && !mkdir($targetDir) && !is_dir($targetDir)) {
+                $errorFilesInfo[] = [
+                    "filename" => $targetDir,
+                    "status"   => gT("Could not create directory")
+                ];
+            }
+            while ($direntry = readdir($directory)) {
```
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/models/services/CopySurveyResources.php:58]: OK : return at line 49

**Comment by Shnoulle** [comment]
And in my opinion : the test must just test ressource copy, you do both. Can keep, but not part of test needed ;)

File: application/models/services/CopySurveyResources.php [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        $copiedFilesInfo = [];
+        $errorFilesInfo = [];
+
+        if (is_dir($sourceDir)) {
+            $directory = opendir($sourceDir);
+            if (!$directory) {
+                $errorFilesInfo[] = [
+                    "filename" => '',
+                    "status"   => gT("Source dir not found - maybe a permission problem?")
```
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/models/services/CopySurveyResources.php:47]: Mean : if `$directory` didn't exist : it's not an error, it's just a survey without files.

[the remark this answers is not in the document]
Comment (gabrieljenik) [reply application/models/services/CopySurveyResources.php:47]: The case would be someone checking "copy resources" but no resources associated.
Will review.

File: application/helpers/admin/import_helper.php [hunk 8]
```diff
@@ -1652,8 +1652,10 @@ function XMLImportSurvey($sFullFilePath, $sXMLdata = null, $sNewSurveyName = nul
...
```
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/helpers/admin/import_helper.php:1655]: - There is no mantis ticket for that. : _right, no exist_
- Not sure it is necessary. : _i really think of issue fixed must have a line in [release_notes.txt](https://github.com/LimeSurvey/LimeSurvey/blob/master/docs/release_notes.txt) : i use it in forum sometimes._

Then : yes just a false commit :)

File: application/models/services/CopySurveyResources.php [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+                ];
+                return [$copiedFilesInfo, $errorFilesInfo];
+            }
+
+            if (!is_dir($targetDir) && !mkdir($targetDir) && !is_dir($targetDir)) {
```
[the remark this answers is not in the document]
Comment (gabrieljenik) [reply application/models/services/CopySurveyResources.php:52]: Not sure I understand. Can you please add more detail? Thanks!

File: application/models/services/CopySurveyResources.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/models/services/CopySurveyResources.php:47]: You don't test is upload/survey/sid exist here : https://github.com/LimeSurvey/LimeSurvey/pull/2206/files#diff-f8c970472fc354612651d57fa98027ecd114f95cf42e4594e0b0b61340ec217fR2207 

Then : you try to copy survey ressource of survey without ressource :)

File: tests/unit/helpers/CopySurveyResourcesTest.php [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        if (!is_dir($destdir)) {
+            $dirCreated = mkdir($destdir, 0777, true);
+        }
+        $this->assertTrue($dirCreated, "Couldn't create dir '$destdir'");
+        $file = self::$dataFolder .'/file_upload/dalahorse.jpg';
+        $this->assertTrue(file_exists($file));
+        copy($file, $destdir . "dalahorse.jpg");
+        $this->assertTrue(file_exists($destdir . "dalahorse.jpg"));
+
+        // Copy survey
```
[the remark this answers is not in the document]
Comment (gabrieljenik) [reply tests/unit/helpers/CopySurveyResourcesTest.php:35]: We need some source files to copy...

**Comment by gabrieljenik** [comment]
> Seems there are a fix , then need to be in commit "Fixed issue : link updated when copy survey in all condition"

Changing commit messages can be really difficult. Do you think it is that necessary?

> Survey without file (no sourcedir) are not an error. I have a lot of surey without related survey/upload dir

Will review

File: tests/unit/helpers/CopySurveyResourcesTest.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (Shnoulle) [reply tests/unit/helpers/CopySurveyResourcesTest.php:35]: I mean the part where you test survey/copyy

```
        \Yii::import('application.helpers.export_helper', true);
        \Yii::import('application.helpers.admin.import_helper', true);
        $sourceData = \surveyGetXMLData($sourceSid);
        $importResults = \XMLImportSurvey('', $sourceData);
        $this->assertIsArray($importResults);
        $this->assertArrayHasKey('newsid', $importResults);
```

just for `$targetSid` ?

File: application/models/services/CopySurveyResources.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/models/services/CopySurveyResources.php:52]: I mean : 

What happen if directory already exist ?  It muts not happen, but maybe if happen : must send a warning to user. 

```
 if (!is_dir($targetDir)) {
     $errorFilesInfo[] = [
        "filename" => '',
        "status"   => gT("Destination dir already exist! Can cpontain more files than source dir.")
    ];
 } elseif(!mkdir($targetDir) && !is_dir($targetDir))) {
     $errorFilesInfo[] = [
        "filename" => '',
        "status"   => gT("Destination dir already exist! Can cpontain more files than source dir.")
    ];
    /* Must return : can not do the copy */
    return [$copiedFilesInfo, $errorFilesInfo];
 }
```

PS : you **must** return if $targetDir can not be created.

[the remark this answers is not in the document]
Comment (Shnoulle) [reply application/models/services/CopySurveyResources.php:52]: fixed : gT("Could not create directory") for the second error.

**Comment by Shnoulle** [comment]
Then finally : update to be done

1. Survey without file (no sourcedir) are not an error. Must not return an error
2. If targetdir can not be created : must return an error (and don't try to copy inside)
3. If targetdir exist : must copy but return a warning
4. _If possible_ : have a Fixed issue commit 

tell me if you're OK with all of this of if you think i misunderstanding some point

**Review by Shnoulle — APPROVED** [review]

**Comment by Shnoulle** [comment]
>  ExcessiveMethodLength  The method copyDirectory() has 63 lines of code. Current threshold is set to 60. Avoid really long methods.

So sorry !

File: application/models/services/CopySurveyResources.php [hunk 11]
```diff
@@ -25,82 +31,93 @@ public function copyResources($sourceSurveyId, $targetSurveyId)
         $sourceDir = \Yii::app()->getConfig('uploaddir') . "/surveys/{$sourceSurveyId}/";
         $targetDir = \Yii::app()->getConfig('uploaddir') . "/surveys/{$targetSurveyId}/";
 
-        // Don't do anything if the source survey doesn't have a directory
-        if (!is_dir($sourceDir)) {
-            return [[], []];
+        // Only try to copy the resources if the source survey has a resources directory
+        if (is_dir($sourceDir)) {
```
Comment (Shnoulle) [inline application/models/services/CopySurveyResources.php:35]: :+1: really better !

**Review by Shnoulle — APPROVED** [review]
:+1:

**Comment by gabrieljenik** [comment]
> tell me if you're OK with all of this of if you think i misunderstanding some point

Thanks!

> So sorry !

It is how it is! :)

**Pull request merged**